File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/generic/GenericRotatableModelBlock.java
package generations.gg.generations.core.generationscore.common.world.level.block.generic;

import dev.architectury.registry.registries.RegistrySupplier;
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.util.MathUtils;
import generations.gg.generations.core.generationscore.common.world.level.block.entities.MutableBlockEntityType;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.*;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.DirectionProperty;
import net.minecraft.world.level.block.state.properties.IntegerProperty;
import net.minecraft.world.level.material.Fluids;
import net.minecraft.world.level.storage.loot.LootParams;
import net.minecraft.world.phys.AABB;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

@SuppressWarnings("deprecation")
public class GenericRotatableModelBlock<T extends BlockEntity & ModelContextProviders.ModelProvider> extends GenericModelBlock<T> {
    private static final Map<Size, Map<Integer, IntegerProperty>> SIZE_PROPERTIES = new HashMap<>();

    public static final DirectionProperty FACING = HorizontalDirectionalBlock.FACING;

    private static final BiFunction<BlockPos, BlockState, BlockPos> DEFAULT_BLOCK_ROTATE_POS_FUNCTION = (pos, state) -> {
        if(state.getBlock() instanceof GenericRotatableModelBlock<?> block) {
            var facing = state.getValue(FACING);
            var x = block.getWidthValue(state);
            var y = block.getHeightValue(state);
            var z = block.getLengthValue(state);

            return block.adjustBlockPos(pos, facing, x, y, z, false);
        }

        return pos;
    };

    protected int width;
    protected int height;
    protected int length;

    protected GenericRotatableModelBlock(Properties materialIn, RegistrySupplier<MutableBlockEntityType<T>> blockEntityFunction, BiFunction<BlockPos, BlockState, BlockPos> baseBlockPosFunction, ResourceLocation model, int width, int height, int length) {
        super(materialIn, blockEntityFunction, baseBlockPosFunction, model);
        assignSize(width, height, length);
        reassignStateDefinition();
        this.registerDefaultState(createDefaultState());

    }

    protected GenericRotatableModelBlock(Properties materialIn, RegistrySupplier<MutableBlockEntityType<T>> blockEntityFunction, BiFunction<BlockPos, BlockState, BlockPos> baseBlockPosFunction, ResourceLocation model) {
        this(materialIn, blockEntityFunction, baseBlockPosFunction, model, 0, 0, 0);
    }

    public GenericRotatableModelBlock(Properties properties, RegistrySupplier<MutableBlockEntityType<T>> blockEntityFunction, ResourceLocation model, int width, int height, int length) {
        super(properties, blockEntityFunction, DEFAULT_BLOCK_ROTATE_POS_FUNCTION, model);
        assignSize(width, height, length);
        reassignStateDefinition();
        this.registerDefaultState(createDefaultState());
    }

    protected GenericRotatableModelBlock(Properties properties, RegistrySupplier<MutableBlockEntityType<T>> blockEntityFunction, ResourceLocation model) {
        this(properties, blockEntityFunction, DEFAULT_BLOCK_ROTATE_POS_FUNCTION, model, 0, 0, 0);
    }

    private void assignSize(int width, int height, int length) {
        this.width = width;
        if (width != 0) SIZE_PROPERTIES.computeIfAbsent(Size.WIDTH, value -> new HashMap<>()).computeIfAbsent(width, value -> IntegerProperty.create("width", 0, value));
        this.height = height;
        if(height != 0) SIZE_PROPERTIES.computeIfAbsent(Size.HEIGHT, value -> new HashMap<>()).computeIfAbsent(height, value -> IntegerProperty.create("height", 0, value));
        this.length = length;
        if(length != 0) SIZE_PROPERTIES.computeIfAbsent(Size.LENGTH, value -> new HashMap<>()).computeIfAbsent(length, value -> IntegerProperty.create("length", 0, value));
    }

    protected void reassignStateDefinition() {
        StateDefinition.Builder<Block, BlockState> builder = new StateDefinition.Builder<>(this);
        this.createBlockStateDefinition(builder);
        this.stateDefinition = builder.create(Block::defaultBlockState, BlockState::new);
    }

    protected BlockState createDefaultState() {
        return setSize(super.createDefaultState().setValue(FACING, Direction.NORTH), getBaseX(), 0, getBaseZ());
    }

    @Override
    public @NotNull BlockState mirror(BlockState state, Mirror mirrorIn) {
        return state.rotate(mirrorIn.getRotation(state.getValue(FACING)));
    }

    @Override
    public @NotNull BlockState rotate(BlockState state, Rotation direction) {
        return state.setValue(FACING, direction.rotate(state.getValue(FACING)));
    }

    @Override
    public BlockState getStateForPlacement(@NotNull BlockPlaceContext context) {
        var pos = context.getClickedPos();
        var level = context.getLevel();
        var dir = context.getHorizontalDirection().getOpposite();

        var state = super.getStateForPlacement(context);

        if(pos.getY() < level.getMaxBuildHeight() - height && isAreaClear(level, dir, pos)) {
            return setSize(state.setValue(FACING, dir), getBaseX(), 0, getBaseZ());
        } else {
            return null;
        }
    }

    public int getBaseX() {
        return 0;
    }

    public int getBaseZ() {
        return 0;
    }

    protected boolean isAreaClear(Level level, Direction dir, BlockPos pos) {

        for (int x = 0; x < width + 1; x++) {
            for (int z = 0; z < length + 1; z++) {
                for (int y = 0; y < height + 1; y++) {
                    if(!validPosition(x,y,z)) continue;

                    var blockPos = adjustBlockPos(pos, dir, x, y, z, true);

                    var state = level.getBlockState(blockPos);

                    if(!state.canBeReplaced() || state.is(this)) {
                        return false;
                    }
                }
            }
        }

        return true;
    }

    protected BlockPos adjustBlockPos(BlockPos pos, Direction dir, int x, int y, int z, boolean positive) {
        return positive ? pos.relative(dir.getCounterClockWise(), adjustX(x)).relative(Direction.UP, y).relative(dir.getOpposite(), adjustZ(z)) : pos.relative(dir.getClockWise(), adjustX(x)).relative(Direction.DOWN, y).relative(dir, adjustZ(z));
    }

    protected boolean validPosition(int x, int y, int z) {
        return true;
    }

    protected boolean matches(BlockState state, int x, int y, int z) {
        return getWidthValue(state) == x && getHeightValue(state) == y && getLengthValue(state) == z;
    }

    @Override
    public boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
        var dir = state.getValue(FACING);
        var base = getBaseBlockPos(pos, state);

        return isAreaClear((Level) level, dir, base);
    }

    @Override
    public void playerWillDestroy(Level level, @NotNull BlockPos pos, @NotNull BlockState state, Player player) {
        pos = getBaseBlockPos(pos, state);
        var facing = state.getValue(FACING);

        for (int x = 0; x <= width; x++) {
            for (int z = 0; z <= length; z++) {
                for (int y = 0; y <= height; y++) {
                    if(!validPosition(x,y, z)) continue;


                    var blockPos = adjustBlockPos(pos, facing, x, y,z, true);

                    level.destroyBlock(blockPos, !player.isCreative() && x == 0 && y == 0 && z == 0);
                }
            }
        }

        super.playerWillDestroy(level, pos, state, player);
    }


    @Override
    public void playerDestroy(Level level, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack tool) {
        super.playerDestroy(level, player, pos, Blocks.AIR.defaultBlockState(), blockEntity, tool);
    }

    @Override
    public @NotNull List<ItemStack> getDrops(@NotNull BlockState state, LootParams.@NotNull Builder params) {
        if(getWidthValue(state) == getBaseX() && getLengthValue(state) == 0 && getHeightValue(state) == getBaseZ()) return super.getDrops(state, params);
        else return Collections.emptyList();
    }

    private int getValue(LevelReader level, BlockPos pos, Direction direction, Size size) {
        var blockState = level.getBlockState(pos.relative(direction));
        return blockState.is(this) ? getValue(blockState, size) : -1;
    }

    public int getValue(BlockState blockState, Size size) {
        return switch (size) {
            case WIDTH -> getWidthValue(blockState);
            case HEIGHT -> getHeightValue(blockState);
            case LENGTH -> getLengthValue(blockState);
        };
    }

    private int getSize(Direction.Axis axis) {
        return switch (axis) {
            case X -> width;
            case Y -> height;
            case Z -> length;
        };
    }

//    @Override
//    public @NotNull BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor level, BlockPos currentPos, BlockPos neighborPos) {
//        if(canSurvive(state, level, currentPos)) return super.updateShape(state, direction, neighborState, level, currentPos, neighborPos);
//        else return Blocks.AIR.defaultBlockState();
//    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.@NotNull Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        builder.add(FACING);
        if (width != 0) builder.add(SIZE_PROPERTIES.get(Size.WIDTH).get(width));
        if (height != 0) builder.add(SIZE_PROPERTIES.get(Size.HEIGHT).get(height));
        if (length != 0) builder.add(SIZE_PROPERTIES.get(Size.LENGTH).get(length));
    }

    @Override
    public AABB computeRenderBoundingBox(Level level, BlockPos pos, BlockState state) {
        return new AABB(pos, adjustBlockPos(getBaseBlockPos(pos, state), state.getValue(FACING), width + 1, height + 1, length + 1, true));
    }

    @Override
    public boolean canRender(Level level, BlockPos blockPos, BlockState blockState) {
        return getWidthValue(blockState) == getBaseX() || getHeightValue(blockState) == 0 || getLengthValue(blockState) == getBaseZ();
    }

    public float getAngle(BlockState state) {
        return state.getValue(FACING).toYRot();
    }

    @Override
    public void setPlacedBy(@NotNull Level level, @NotNull BlockPos pos, @NotNull BlockState state, @Nullable LivingEntity placer, @NotNull ItemStack stack) {
        for (int x = 0; x <= width; x++) {
            for (int z = 0; z <= length; z++) {
                for (int y = 0; y <= height; y++) {
                    if(!validPosition(x,y, z)) continue;

                    if(x == getBaseX() && y == 0 && z == getBaseZ()) continue;
                    var blockPos = adjustBlockPos(pos, state.getValue(FACING), x, y, z, true);
                    level.setBlock(blockPos, setSize(state.setValue(WATERLOGGED, level.getFluidState(blockPos).getType() == Fluids.WATER), x, y, z), 2);
                }
            }
        }
    }

    public int getLengthValue(BlockState state) {
        return state.is(this) && length != 0 ? state.getValue(getLengthProperty()) : Integer.valueOf(0);
    }

    public BlockState setLengthValue(BlockState state, int value) {
        return state.is(this) && length > 0 && MathUtils.between(value, 0, length) ? state.setValue(getLengthProperty(), value) : state;
    }

    public int getHeightValue(BlockState state) {
        return state.is(this) && height != 0 ? state.getValue(getHeightProperty()) : Integer.valueOf(0);
    }

    public BlockState setHeighthValue(BlockState state, int value) {
        return state.is(this) && height > 0 && MathUtils.between(value, 0, height) ? state.setValue(getHeightProperty(), value) : state;
    }

    public int getWidthValue(BlockState state) {
        return state.is(this) && width != 0 ? state.getValue(getWidthProperty()) : Integer.valueOf(0);
    }

    public BlockState setWidthValue(BlockState state, int value) {
        return state.is(this) && width > 0 && MathUtils.between(value, 0, width) ? state.setValue(getWidthProperty(), value) : state;
    }

    public IntegerProperty getWidthProperty() {
        return SIZE_PROPERTIES.get(Size.WIDTH).get(width);
    }

    public IntegerProperty getHeightProperty() {
        return SIZE_PROPERTIES.get(Size.HEIGHT).get(height);
    }

    public IntegerProperty getLengthProperty() {
        return SIZE_PROPERTIES.get(Size.LENGTH).get(length);
    }

    public BlockState setSize(BlockState state, int x, int y, int z) {
        state = setWidthValue(state, x);
        state = setHeighthValue(state, y);
        state = setLengthValue(state, z);
        return state;
    }

    public int width() {
        return width;
    }

    public int height() {
        return height;
    }

    public int length() {
        return length;
    }

    public int adjustX(int x) {
        return x - getBaseX();
    }

    public int adjustZ(int z) {
        return z - getBaseZ();
    }

    public enum Size {//TODO: THink of better name
        WIDTH, LENGTH, HEIGHT
    }

    public boolean shouldRotateSpecial() {
        return true;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/generic/GenericRotatableModelBlock.kt
package generations.gg.generations.core.generationscore.common.world.level.block.generic

import dev.architectury.registry.registries.RegistrySupplier
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders
import generations.gg.generations.core.generationscore.common.util.extensions.between
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.block.*
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.properties.DirectionProperty
import net.minecraft.world.level.block.state.properties.IntegerProperty
import net.minecraft.world.level.material.Fluids
import net.minecraft.world.level.storage.loot.LootParams
import net.minecraft.world.phys.AABB

abstract class GenericRotatableModelBlock<T>(properties: Properties, blockEntityFunction: RegistrySupplier<BlockEntityType<T>>, posFunction: (BlockPos, BlockState) -> BlockPos = DEFAULT_BLOCK_ROTATE_POS_FUNCTION, model: ResourceLocation? = null, val width: Int = 0, val height: Int = 0, val length: Int = 0) : GenericModelBlock<T>(properties, blockEntityFunction, posFunction, model) where T : BlockEntity, T : ModelContextProviders.ModelProvider {

    init {
        assignSize(width, height, length)
        reassignStateDefinition()
        this.registerDefaultState(createDefaultState())

        super.modelResource
    }

    private fun assignSize(width: Int, height: Int, length: Int) {
        val function: (Size, Int) -> Unit = { size, value -> if(value != 0) SIZE_PROPERTIES.computeIfAbsent(size) { mutableMapOf() }.computeIfAbsent(value) { IntegerProperty.create(size.name.lowercase(), 0, value) } }

        function.invoke(Size.WIDTH, width)
        function.invoke(Size.HEIGHT, height)
        function.invoke(Size.LENGTH, length)
    }

    protected fun reassignStateDefinition() {
        val builder = StateDefinition.Builder<Block, BlockState>(this)
        this.createBlockStateDefinition(builder)
        this.stateDefinition = builder.create(Block::defaultBlockState, ::BlockState)
    }

    override fun createDefaultState(): BlockState {
        return setSize(
            super.createDefaultState().setValue(FACING, Direction.NORTH),
            baseX, 0,
            baseZ
        )
    }

    public override fun mirror(state: BlockState, mirrorIn: Mirror): BlockState = state.rotate(mirrorIn.getRotation(state.getValue(FACING)))

    public override fun rotate(state: BlockState, direction: Rotation): BlockState = state.setValue(FACING, direction.rotate(state.getValue(FACING)))

    override fun getStateForPlacement(context: BlockPlaceContext): BlockState? {
        val pos: BlockPos = context.clickedPos
        val level: Level = context.level
        val dir: Direction = context.horizontalDirection.opposite

        return super.getStateForPlacement(context)?.takeIf { pos.y < level.maxBuildHeight - height && isAreaClear(level, dir, pos) }?.let {
            setSize(
                it.setValue(FACING, dir),
                baseX, 0,
                baseZ
            )
        }
    }

    open val baseX: Int
        get() = 0

    val baseZ: Int
        get() = 0

    private fun isAreaClear(level: LevelReader, dir: Direction, pos: BlockPos): Boolean {
        for (x in 0 until width + 1) {
            for (z in 0 until length + 1) {
                for (y in 0 until height + 1) {
                    if (!validPosition(x, y, z)) continue

                    val blockPos: BlockPos = adjustBlockPos(pos, dir, x, y, z, true)

                    val state: BlockState = level.getBlockState(blockPos)

                    if (!state.canBeReplaced() || state.`is`(this)) {
                        return false
                    }
                }
            }
        }

        return true
    }

    protected fun adjustBlockPos(pos: BlockPos, dir: Direction, x: Int, y: Int, z: Int, positive: Boolean): BlockPos {
        return if (positive) pos.relative(dir.counterClockWise, adjustX(x)).relative(Direction.UP, y)
            .relative(dir.opposite, adjustZ(z)) else pos.relative(dir.clockWise, adjustX(x)).relative(
            Direction.DOWN, y
        ).relative(dir, adjustZ(z))
    }

    protected open fun validPosition(x: Int, y: Int, z: Int): Boolean {
        return true
    }

    protected fun matches(state: BlockState, x: Int, y: Int, z: Int): Boolean {
        return getWidthValue(state) == x && getHeightValue(state) == y && getLengthValue(state) == z
    }

    public override fun canSurvive(state: BlockState, level: LevelReader, pos: BlockPos): Boolean {
        val dir: Direction = state.getValue(FACING)
        val base: BlockPos = getBaseBlockPos(pos, state)

        return isAreaClear(level, dir, base)
    }

    override fun playerWillDestroy(level: Level, pos: BlockPos, state: BlockState, player: Player): BlockState {
        var position: BlockPos = getBaseBlockPos(pos, state)
        val facing: Direction = state.getValue(FACING)

        for (x in 0..width) {
            for (z in 0..length) {
                for (y in 0..height) {
                    if (!validPosition(x, y, z)) continue


                    val blockPos: BlockPos = adjustBlockPos(position, facing, x, y, z, true)



                    level.destroyBlock(blockPos, !player.isCreative && x == baseX && y == 0 && z == baseZ)
                }
            }
        }

        return super.playerWillDestroy(level, position, state, player)
    }


    override fun playerDestroy(
        level: Level,
        player: Player,
        pos: BlockPos,
        state: BlockState,
        blockEntity: BlockEntity?,
        tool: ItemStack
    ) {
        super.playerDestroy(level, player, pos, Blocks.AIR.defaultBlockState(), blockEntity, tool)
    }

    public override fun getDrops(state: BlockState, params: LootParams.Builder): List<ItemStack> {
        if (getWidthValue(state) == baseX && getLengthValue(state) == 0 && getHeightValue(state) == baseZ) return super.getDrops(
            state,
            params
        )
        else return emptyList()
    }

    private fun getValue(level: LevelReader, pos: BlockPos, direction: Direction, size: Size): Int {
        val blockState: BlockState = level.getBlockState(pos.relative(direction))
        return if (blockState.`is`(this)) getValue(blockState, size) else -1
    }

    fun getValue(blockState: BlockState, size: Size): Int {
        return when (size) {
            Size.WIDTH -> getWidthValue(blockState)
            Size.HEIGHT -> getHeightValue(blockState)
            Size.LENGTH -> getLengthValue(blockState)
        }
    }

    fun getSizeProperty(size: Size): IntegerProperty? = when(size) {
        Size.WIDTH -> widthProperty
        Size.LENGTH -> lengthProperty
        Size.HEIGHT -> heightProperty
    }

    private fun getSize(axis: Direction.Axis): Int {
        return when (axis) {
            Direction.Axis.X -> width
            Direction.Axis.Y -> height
            Direction.Axis.Z -> length
        }
    }

    //    @Override
    //    public @NotNull BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor level, BlockPos currentPos, BlockPos neighborPos) {
    //        if(canSurvive(state, level, currentPos)) return super.updateShape(state, direction, neighborState, level, currentPos, neighborPos);
    //        else return Blocks.AIR.defaultBlockState();
    //    }
    override fun createBlockStateDefinition(builder: StateDefinition.Builder<Block, BlockState>) {
        super.createBlockStateDefinition(builder)
        builder.add(FACING)
        if (width != 0) builder.add(
            SIZE_PROPERTIES.get(Size.WIDTH)!!
                .get(width)
        )
        if (height != 0) builder.add(
            SIZE_PROPERTIES.get(Size.HEIGHT)!!
                .get(height)
        )
        if (length != 0) builder.add(
            SIZE_PROPERTIES.get(Size.LENGTH)!!
                .get(length)
        )
    }

    override fun computeRenderBoundingBox(level: Level, pos: BlockPos, state: BlockState): AABB {
        return AABB.encapsulatingFullBlocks(
            pos, adjustBlockPos(
                getBaseBlockPos(pos, state), state.getValue(
                    FACING
                ), width + 1, height + 1, length + 1, true
            )
        )
    }

    override fun canRender(level: Level, blockPos: BlockPos, blockState: BlockState): Boolean {
        return getWidthValue(blockState) == baseX && getHeightValue(blockState) == 0 && getLengthValue(blockState) == baseZ
    }

    fun getAngle(state: BlockState): Float {
        return state.getValue(FACING).toYRot()
    }

    override fun setPlacedBy(level: Level, pos: BlockPos, state: BlockState, placer: LivingEntity?, stack: ItemStack) {
        for (x in 0..width) {
            for (z in 0..length) {
                for (y in 0..height) {
                    if (!validPosition(x, y, z)) continue

                    if (x == baseX && y == 0 && z == baseZ) continue
                    val blockPos: BlockPos = adjustBlockPos(pos, state.getValue(FACING), x, y, z, true)
                    level.setBlock(
                        blockPos,
                        setSize(
                            state.setValue(WATERLOGGED, level.getFluidState(blockPos).type === Fluids.WATER),
                            x,
                            y,
                            z
                        ),
                        2
                    )
                }
            }
        }
    }

    fun getLengthValue(state: BlockState): Int {
        return if (state.`is`(this) && length != 0) state.getValue(lengthProperty) else 0
    }

    fun setLengthValue(state: BlockState, value: Int): BlockState {
        return if (state.`is`(this) && length > 0 && value.between(0, length)) state.setValue(
            lengthProperty, value
        ) else state
    }

    fun getHeightValue(state: BlockState): Int {
        return if (state.`is`(this) && height != 0) state.getValue(heightProperty) else 0
    }

    fun setHeighthValue(state: BlockState, value: Int): BlockState {
        return if (state.`is`(this) && height > 0 && value.between(0, height)) state.setValue(
            heightProperty, value
        ) else state
    }

    fun getWidthValue(state: BlockState): Int {
        return if (state.`is`(this) && width != 0) state.getValue(widthProperty) else 0
    }

    fun setWidthValue(state: BlockState, value: Int): BlockState {
        return if (state.`is`(this) && width > 0 && value.between(0, width)) state.setValue(
            widthProperty, value
        ) else state
    }

    val widthProperty: IntegerProperty?
        get() = SIZE_PROPERTIES[Size.WIDTH]?.get(width)

    val heightProperty: IntegerProperty?
        get() = SIZE_PROPERTIES[Size.HEIGHT]?.get(height)

    val lengthProperty: IntegerProperty?
        get() = SIZE_PROPERTIES[Size.LENGTH]?.get(length)

    fun setSize(state: BlockState, x: Int, y: Int, z: Int): BlockState {
        var state: BlockState = state
        state = setWidthValue(state, x)
        state = setHeighthValue(state, y)
        state = setLengthValue(state, z)
        return state
    }

    fun adjustX(x: Int): Int {
        return x - baseX
    }

    fun adjustZ(z: Int): Int {
        return z - baseZ
    }

    enum class Size {
        //TODO: THink of better name
        WIDTH, LENGTH, HEIGHT
    }

    open fun shouldRotateSpecial(): Boolean {
        return true
    }

    public override fun getSoundType(state: BlockState): SoundType {
        return if ((getWidthValue(state) == baseX && getHeightValue(state) == 0 && getLengthValue(state) == baseZ)) super.getSoundType(
            state
        ) else SoundType.EMPTY
    }

    companion object {
        private val SIZE_PROPERTIES: MutableMap<Size, MutableMap<Int, IntegerProperty>> = mutableMapOf()

        @JvmField
        val FACING: DirectionProperty = HorizontalDirectionalBlock.FACING

        private val DEFAULT_BLOCK_ROTATE_POS_FUNCTION: (BlockPos, BlockState) -> BlockPos = { pos, state ->
            val block = state.block
                if (block is GenericRotatableModelBlock<*>) {
                    val facing: Direction = state.getValue(FACING)
                    val x: Int = block.getWidthValue(state)
                    val y: Int = block.getHeightValue(state)
                    val z: Int = block.getLengthValue(state)

                    block.adjustBlockPos(pos, facing, x, y, z, false)
                } else pos
            }
    }
}